src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs

```diff
@@ -115,34 +140,93 @@ internal async Task<IEnumerable<IProjectVersionedValue<RestoreData>>> RestoreAsy
                 return true;
+            }
 
             _latestHash = hash;
 
+            _nugetRestoreStarted = true;
             JoinableTask<bool> joinableTask = JoinableFactory.RunAsync(() =>
             {
-                return NominateForRestoreAsync(restoreInfo, _projectAsynchronousTasksService.UnloadCancellationToken);
+                return NominateForRestoreAsync(restoreInfo!, _projectAsynchronousTasksService.UnloadCancellationToken);
             });
 
+            SaveNominatedConfiguredVersions(value.ConfiguredInputs);
+
             _projectAsynchronousTasksService.RegisterAsyncTask(joinableTask,
                                                                ProjectCriticalOperation.Build | ProjectCriticalOperation.Unload | ProjectCriticalOperation.Rename,
                                                                registerFaultHandler: true);
 
             // Prevent overlap until Restore completes
             bool success = await joinableTask;
 
-            HintProjectDependentFile(restoreInfo);
+            _nugetRestoreStarted = false;
+
+            HintProjectDependentFile(restoreInfo!);
 
             return success;
         }
 
+        private bool IsProjectConfigurationVersionOutOfDate(IReadOnlyCollection<PackageRestoreConfiguredInput> configuredInputs)
+        {
+            Assumes.Present(_project.Services.ActiveConfiguredProjectProvider);
+            Assumes.Present(_project.Services.ActiveConfiguredProjectProvider.ActiveConfiguredProject);
+
+            var activeConfiguredProject = _project.Services.ActiveConfiguredProjectProvider.ActiveConfiguredProject;
+
+            if (!_savedNominatedConfiguredVersion.TryGetValue(activeConfiguredProject.ProjectConfiguration, out IComparable savedVersionOfActiveConfiguration))
+            {
+                return false;
+            }
+
+            // Nuget only cares about the active configuration
+            foreach (var configuredInput in configuredInputs)
+            {
+                if (configuredInput.ProjectConfiguration.Equals(activeConfiguredProject.ProjectConfiguration) &&
+                    configuredInput.ConfiguredProjectVersion.IsLaterThanOrEqualTo(savedVersionOfActiveConfiguration))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private void SaveNominatedConfiguredVersions(IReadOnlyCollection<PackageRestoreConfiguredInput> configuredInputs)
+        {
+            lock (SyncObject)
+            {
+                _savedNominatedConfiguredVersion.Clear();
+
+                foreach (var configuredInput in configuredInputs)
+                {
+                    _savedNominatedConfiguredVersion[configuredInput.ProjectConfiguration] = configuredInput.ConfiguredProjectVersion;
+                }
+
+                if (_whenNominatedTask is not null)
+                {
+                    _whenNominatedTask.SetResult(true);
+                }
```

Comment:
you need set _whenNominatedTask to null, because the project might become dirty again in the future. In the new cycle, you will have to create a new completion source.